test/edgehog_device_forwarder/tokens_test.exs

```diff
@@ -0,0 +1,44 @@
+# Copyright 2023 SECO Mind Srl
+# SPDX-License-Identifier: Apache-2.0
+
+defmodule EdgehogDeviceForwarder.TokensTests do
+  alias EdgehogDeviceForwarder.Tokens
+  use EdgehogDeviceForwarder.CacheCase
+
+  setup %{valid_token: token, socket: socket} do
+    :ok = Tokens.reserve(token)
+    :ok = Tokens.register(token, socket)
+
+    :ok
+  end
+
+  test "register/2 saves the socket in cache", %{socket: socket} do
+    new_token = "some_other_token"
+    assert {:error, :token_not_found} == Tokens.fetch(new_token)
+    assert :ok == Tokens.reserve(new_token)
+    assert :ok == Tokens.register(new_token, socket)
+    assert {:ok, socket} == Tokens.fetch(new_token)
+  end
+
+  test "close/1 removes the token from the cache", %{valid_token: token} do
+    assert :ok == Tokens.close(token)
+    assert {:error, :token_not_found} == Tokens.fetch(token)
+  end
+
+  describe "add_device_info/2" do
+    test "adds device info to existing token", %{valid_token: token} do
+      realm = "some_realm"
+      device_id = "some_device_id"
+      assert :ok == Tokens.add_device_info(token, device_id, realm)
```

Comment:
I would also add a line checking that `realm` and `device_id` are as expected:
```elixir
assert %Data{realm: ^realm, device_id: ^device_id} = Tokens.fetch(token)
```

Reply:
hmmm
at the moment `fetch` only returns the socket

but it would be better to have a `fetch_device_socket` for that, and have `fetch` return the `Data` struct